src/basics.ml

```diff
@@ -239,3 +343,52 @@ let term_of_rhs : rule -> term = fun r ->
     TE_Some(Bindlib.unbox (Bindlib.bind_mvar vars p))
   in
   Bindlib.msubst r.rhs (Array.mapi fn r.vars)
+
+(** [to_terms r] translates the rule [r] into a pair of terms. The pattern
+    variables in the LHS are replaced by metavariables and the terms with
+    environment in the RHS are replaced by their corresponding metavariables.
+    *)
+let to_terms : sym * rule -> term * term = fun (s, r) ->
+  let arity = Bindlib.mbinder_arity r.rhs in
+  let metas = Array.init arity (fun _ -> None) in
+  let lhs = List.map (fun p -> Bindlib.unbox (to_m 0 metas p)) r.lhs in
+  let lhs = add_args (symb s) lhs in
+  (* [to_term_env m] computes the term with environment corresponding to the
+     metavariable [m]. *)
+  let to_term_env : meta option -> term_env = fun m ->
+    let m = match m with Some m -> m | None -> assert false in
+    let xs = Array.init m.meta_arity (Printf.sprintf "x%i") in
+    let xs = Bindlib.new_mvar mkfree xs in
+    let ar = Array.map _Vari xs in
+    TE_Some (Bindlib.unbox (Bindlib.bind_mvar xs (_Meta m ar))) in
+  let terms_env = Array.map to_term_env metas in
+  let rhs = Bindlib.msubst r.rhs terms_env in
+  (lhs, rhs)
+
+(** [to_closed_terms r] translates the rule [r] into a pair of terms. The
+    pattern variables in the LHS are replaced by fresh symbols as in the
+    function [to_closed] and the terms with environment in the RHS are
+    replaced by their corresponding symbols. *)
+let to_closed_terms : sym * rule -> term * term = fun (s, r) ->
+  let arity = Bindlib.mbinder_arity r.rhs in
+  let symbs = Array.init arity (fun _ -> None) in
+  let lhs = List.map (fun p -> Bindlib.unbox (to_closed symbs p)) r.lhs in
+  let lhs = add_args (symb s) lhs in
+  let to_term_env : sym option -> term_env = fun s ->
+    let s = match s with Some s -> s | None -> assert false in
+    TE_Some (Bindlib.unbox (Bindlib.bind_mvar [||] (_Symb s Nothing))) in
+  let terms_env = Array.map to_term_env symbs in
+  let rhs = Bindlib.msubst r.rhs terms_env in
+  (lhs, rhs)
+
+(** [check_fo t] checks that [t] is a first-order term. *)
+let rec check_fo : term -> unit = fun t ->
```

Comment:
When a function may raise an exception, this must be said in the comment.
Replace fo by first_order.
Why raising an exception and not returning a boolean? This should be explained.

Reply:
The name is misleading or the code is wrong: this function does not check that the term is first-order. You should specify what is first-order. Why not using iter to implement it?

Reply:
In the algorithm for checking SR, we only consider left-algebraic rule. However, the type inference works for any term, so I left the possibility of applying the algorithm to any rule with LHS containing no metavariable of arity > 0 but the algorithm might not work in this case.

Reply:
Ok, give a more informative name then.